pkg/clients/bigquery.go

```diff
@@ -0,0 +1,242 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package clients encapsulates all clients and associated helper methods to
+// interact with other services.
+package clients
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"cloud.google.com/go/bigquery"
+	"github.com/abcxyz/pkg/logging"
+	"go.uber.org/zap"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+)
+
+// BigQuery provides a client and dataset identifiers.
+type BigQuery struct {
+	projectID string
+	datasetID string
+	client    *bigquery.Client
+	logger    *zap.SugaredLogger
+}
+
+// CheckpointEntry is the shape of an entry to the checkpoint table.
+type CheckpointEntry struct {
+	deliveryID string
+	createdAt  string
+}
+
+// FailureEventEntry is the shape of an entry to the failure_events table.
+type FailureEventEntry struct {
+	deliveryID string
+	createdAt  string
+}
+
+// NewBigQuery creates a new instance of a BigQuery client.
+func NewBigQuery(ctx context.Context, projectID, datasetID string, opts ...option.ClientOption) (*BigQuery, error) {
+	client, err := bigquery.NewClient(ctx, projectID, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new bigquery client: %w", err)
+	}
+
+	return &BigQuery{
+		projectID: projectID,
+		datasetID: datasetID,
+		client:    client,
+		logger:    logging.FromContext(ctx).Named("bigquery"),
+	}, nil
+}
+
+// Close releases any resources held by the BigQuery client.
+func (bq *BigQuery) Close() error {
+	if err := bq.client.Close(); err != nil {
+		return fmt.Errorf("failed to close bigquery client: %w", err)
+	}
+	return nil
+}
+
+// Check if an entry with a given delivery_id already exists in the events
+// table, this attempts to prevent duplicate processing of events. This is used
+// by the webhook service.
+func (bq *BigQuery) DeliveryEventExists(ctx context.Context, eventsTableID, deliveryID string) (bool, error) {
+	res, err := bq.makeCountQuery(ctx, eventsTableID, deliveryID)
+	if err != nil {
+		return false, fmt.Errorf("failed to execute DeliveryEventExists: %w", err)
+	}
+
+	return res > 0, nil
+}
+
+// Check if the number of entries with a given delivery_id in the failure-events
+// table exceeds the retry limit. This is used by the webhook service.
+func (bq *BigQuery) FailureEventsExceedsRetryLimit(ctx context.Context, failureEventTableID, deliveryID string, maxRetry int) (bool, error) {
+	res, err := bq.makeCountQuery(ctx, failureEventTableID, deliveryID)
+	if err != nil {
+		return false, fmt.Errorf("failed to execute FailureEventsExceedsRetryLimit: %w", err)
+	}
+
+	bq.logger.Infow(
+		fmt.Sprintf("retry limit check on table %s for deliveryID %s", failureEventTableID, deliveryID),
+		"attempts", res, "deliveryID", deliveryID, "limit", maxRetry,
+	)
```

Comment:
I would suggest only logging this for unexpected things:

```suggestion
if res > 0 && res < maxRetry {
	bq.logger.Debugw("found retries, but does not exceed maxRetries",
		"attempts", res,
		"delivery_id": deliveryID,
		"limit": maxRetry,
	)
}
```